File: ios/shift/shift/UI/Confirmation/ConfirmationViewModel.swift
import Foundation
import Observation
import SwiftUI
import UIKit

struct EditingState: Equatable {
    var weekIndex: Int
    var dayIndex: Int
    var draft: ScheduleDay
}

@MainActor
@Observable
final class ConfirmationViewModel {
    private let repository: any ScheduleRepository
    private let preferencesRepository: any UserPreferencesRepository

    var weeks: [ScheduleWeek]
    var image: UIImage?
    var editing: EditingState?
    var conflictCount = 0
    var showSkipPrompt = false
    var savedSuccessfully = false
    var savedWeeksCount = 0
    var isCancelled = false
    var toastMessage: String?

    init(
        weeks: [ScheduleWeek],
        image: UIImage?,
        repository: any ScheduleRepository,
        preferencesRepository: any UserPreferencesRepository
    ) {
        self.weeks = weeks
        self.image = image
        self.repository = repository
        self.preferencesRepository = preferencesRepository
    }

    func confirm() {
        Task { await runConfirm() }
    }

    private func runConfirm() async {
        var conflicts = 0
        for week in weeks {
            conflicts += (try? await repository.getWeekByDate(week.weekStartDate)) != nil ? 1 : 0
        }
        conflictCount = conflicts
        if conflicts == 0 { await saveAllAndFinish() }
    }

    func proceedWithReplace() {
        Task {
            do {
                for week in weeks {
                    if (try? await repository.getWeekByDate(week.weekStartDate)) != nil {
                        try await repository.replaceWeek(week)
                    } else {
                        try await repository.saveWeek(week)
                    }
                }
                conflictCount = 0
                await finishAfterSave()
            } catch {
                conflictCount = 0
                showToast(saveErrorMessage(error))
            }
        }
    }

    func dismissConflict() { conflictCount = 0 }

    private func saveAllAndFinish() async {
        do {
            for week in weeks { try await repository.saveWeek(week) }
            await finishAfterSave()
        } catch {
            showToast(saveErrorMessage(error))
        }
    }

    func showToast(_ message: String) {
        toastMessage = message
        Task {
            try? await Task.sleep(for: .seconds(3))
            withAnimation { self.toastMessage = nil }
        }
    }

    private func saveErrorMessage(_ error: Error) -> String {
        let nsError = error as NSError
        if nsError.domain == NSPOSIXErrorDomain && nsError.code == 28 {
            return "저장 공간이 부족합니다"
        }
        return "저장에 실패했어요. 다시 시도해주세요."
    }

    private func finishAfterSave() async {
        if !preferencesRepository.isSkipConfirmPromptShown() {
            showSkipPrompt = true
        } else {
            markSaved()
        }
    }

    func answerSkipPrompt(skipInFuture: Bool) {
        preferencesRepository.setSkipConfirm(skipInFuture)
        preferencesRepository.setSkipConfirmPromptShown(true)
        markSaved()
    }

    func cancel() { isCancelled = true }

    func startEdit(weekIndex: Int, dayIndex: Int) {
        guard weekIndex < weeks.count, dayIndex < weeks[weekIndex].days.count else { return }
        let day = weeks[weekIndex].days[dayIndex]
        editing = EditingState(weekIndex: weekIndex, dayIndex: dayIndex, draft: day)
    }

    func updateDraft(_ draft: ScheduleDay) {
        editing?.draft = draft
    }

    func commitEdit() {
        guard let editing else { return }
        var updated = weeks
        var days = updated[editing.weekIndex].days
        days[editing.dayIndex] = editing.draft
        updated[editing.weekIndex] = ScheduleWeek(
            weekStartDate: weeks[editing.weekIndex].weekStartDate, days: days
        )
        weeks = updated
        self.editing = nil
    }

    func dismissEdit() { editing = nil }

    private func markSaved() {
        savedWeeksCount = weeks.count
        savedSuccessfully = true
    }
}

File: android/ui/src/main/kotlin/com/schedule/shift/ui/confirmation/ConfirmationViewModel.kt
package com.schedule.shift.ui.confirmation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.schedule.shift.domain.analytics.AnalyticsEvent
import com.schedule.shift.domain.analytics.AnalyticsTracker
import com.schedule.shift.domain.model.ScheduleDay
import com.schedule.shift.domain.model.ScheduleWeek
import com.schedule.shift.domain.preferences.UserPreferencesRepository
import com.schedule.shift.domain.repository.ScheduleRepository
import com.schedule.shift.domain.widget.WidgetRefresher
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

private const val DAYS_IN_WEEK = 7

private data class EditFieldChange(val field: String, val parsed: String, val corrected: String)

@Suppress("LongParameterList")
@HiltViewModel(assistedFactory = ConfirmationViewModel.Factory::class)
class ConfirmationViewModel @AssistedInject constructor(
    @Assisted private val initialWeeks: List<ScheduleWeek>,
    @Assisted("imageUri") private val imageUri: String?,
    @Assisted("sessionId") private val sessionId: String,
    private val repository: ScheduleRepository,
    private val widgetRefresher: WidgetRefresher,
    private val preferences: UserPreferencesRepository,
    private val tracker: AnalyticsTracker,
) : ViewModel() {

    private val _uiState = MutableStateFlow<ConfirmationUiState>(
        ConfirmationUiState.Reviewing(weeks = initialWeeks, imageUri = imageUri),
    )
    val uiState: StateFlow<ConfirmationUiState> = _uiState.asStateFlow()

    private val confirmStartMs = System.currentTimeMillis()
    private val editedRowIndices = mutableSetOf<Int>()
    private var wasReplace = false

    init {
        tracker.track(AnalyticsEvent.ConfirmShown(sessionId = sessionId, skipped = false))
    }

    fun confirm() {
        val state = _uiState.value as? ConfirmationUiState.Reviewing ?: return
        viewModelScope.launch {
            val conflicts = state.weeks.count { repository.getWeekByDate(it.weekStartDate) != null }
            if (conflicts > 0) {
                _uiState.value = state.copy(conflictCount = conflicts)
                return@launch
            }
            saveAllAndFinish(state.weeks)
        }
    }

    fun proceedWithReplace() {
        val state = _uiState.value as? ConfirmationUiState.Reviewing ?: return
        wasReplace = true
        viewModelScope.launch {
            state.weeks.forEach { week ->
                if (repository.getWeekByDate(week.weekStartDate) != null) {
                    repository.replaceWeek(week)
                } else {
                    repository.saveWeek(week)
                }
            }
            widgetRefresher.refreshAll()
            finishAfterSave()
        }
    }

    fun dismissConflict() {
        val state = _uiState.value as? ConfirmationUiState.Reviewing ?: return
        _uiState.value = state.copy(conflictCount = 0)
    }

    private suspend fun saveAllAndFinish(weeks: List<ScheduleWeek>) {
        weeks.forEach { repository.saveWeek(it) }
        widgetRefresher.refreshAll()
        finishAfterSave()
    }

    private suspend fun finishAfterSave() {
        tracker.track(
            AnalyticsEvent.RegisterComplete(
                sessionId = sessionId,
                editedRows = editedRowIndices.size,
                manualRows = 0,
                replace = wasReplace,
                totalDurationMs = System.currentTimeMillis() - confirmStartMs,
            ),
        )
        val state = _uiState.value as? ConfirmationUiState.Reviewing ?: return
        if (!preferences.isSkipConfirmPromptShown()) {
            _uiState.value = state.copy(showSkipPrompt = true)
        } else {
            _uiState.value = ConfirmationUiState.Saved
        }
    }

    fun answerSkipPrompt(skipInFuture: Boolean) {
        viewModelScope.launch {
            preferences.setSkipConfirm(skipInFuture)
            preferences.setSkipConfirmPromptShown(true)
            _uiState.value = ConfirmationUiState.Saved
        }
    }

    fun cancel() {
        tracker.track(AnalyticsEvent.RegisterAbandon(sessionId = sessionId, lastStep = AnalyticsEvent.RegisterAbandon.LAST_STEP_CONFIRM))
        _uiState.value = ConfirmationUiState.Cancelled
    }

    fun startEdit(weekIndex: Int, dayIndex: Int) {
        val state = _uiState.value as? ConfirmationUiState.Reviewing ?: return
        val day = state.weeks.getOrNull(weekIndex)?.days?.getOrNull(dayIndex) ?: return
        _uiState.value = state.copy(editing = EditingState(weekIndex, dayIndex, day))
    }

    fun updateDraft(draft: ScheduleDay) {
        val state = _uiState.value as? ConfirmationUiState.Reviewing ?: return
        val editing = state.editing ?: return
        _uiState.value = state.copy(editing = editing.copy(draft = draft))
    }

    @Suppress("ReturnCount")
    fun commitEdit() {
        val state = _uiState.value as? ConfirmationUiState.Reviewing ?: return
        val editing = state.editing ?: return
        val originalDay = state.weeks.getOrNull(editing.weekIndex)?.days?.getOrNull(editing.dayIndex) ?: return
        val draft = editing.draft

        if (originalDay != draft) {
            val rowIndex = editing.weekIndex * DAYS_IN_WEEK + editing.dayIndex
            editedRowIndices.add(rowIndex)
            trackChangedFields(originalDay, draft, rowIndex)
        }

        val updatedWeeks = state.weeks.toMutableList()
        val week = updatedWeeks[editing.weekIndex]
        val updatedDays = week.days.toMutableList()
        updatedDays[editing.dayIndex] = draft
        updatedWeeks[editing.weekIndex] = ScheduleWeek(weekStartDate = week.weekStartDate, days = updatedDays)
        _uiState.value = state.copy(weeks = updatedWeeks, editing = null)
    }

    @Suppress("LongParameterList")
    private fun trackChangedFields(original: ScheduleDay, draft: ScheduleDay, rowIndex: Int) {
        buildChangedFields(original, draft).forEach { change ->
            tracker.track(
                AnalyticsEvent.UserEdit(
                    sessionId = sessionId,
                    rowIndex = rowIndex,
                    field = change.field,
                    parsedValue = change.parsed,
                    correctedValue = change.corrected,
                    wasFailedRow = false,
                    editSource = AnalyticsEvent.UserEdit.EDIT_SOURCE_MANUAL,
                ),
            )
        }
    }

    private fun buildChangedFields(original: ScheduleDay, draft: ScheduleDay): List<EditFieldChange> = buildList {
        if (original.startTime != draft.startTime) {
            add(EditFieldChange(AnalyticsEvent.UserEdit.FIELD_START_TIME, original.startTime?.toString() ?: "", draft.startTime?.toString() ?: ""))
        }
        if (original.endTime != draft.endTime) {
            add(EditFieldChange(AnalyticsEvent.UserEdit.FIELD_END_TIME, original.endTime?.toString() ?: "", draft.endTime?.toString() ?: ""))
        }
        if (original.codeLabel != draft.codeLabel) {
            add(EditFieldChange(AnalyticsEvent.UserEdit.FIELD_CODE_LABEL, original.codeLabel, draft.codeLabel))
        }
    }

    fun dismissEdit() {
        val state = _uiState.value as? ConfirmationUiState.Reviewing ?: return
        _uiState.value = state.copy(editing = null)
    }

    @AssistedFactory
    interface Factory {
        @Suppress("LongParameterList")
        fun create(
            weeks: List<ScheduleWeek>,
            @Assisted("imageUri") imageUri: String?,
            @Assisted("sessionId") sessionId: String,
        ): ConfirmationViewModel
    }
}
